pkg/sampling/randomness.go

```diff
@@ -0,0 +1,89 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package sampling // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/sampling"
+
+import (
+	"encoding/binary"
+	"errors"
+	"strconv"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+// numRandomnessValues equals MaxAdjustedCount--this variable has been
+// introduced to improve readability.  Recall that MaxAdjustedCount is
+// 2**56 which is one greater than the maximum RValue
+// ("ffffffffffffff", i.e., "100000000000000").
+const numRandomnessValues = MaxAdjustedCount
+
+// ErrRValueSize is returned by RValueToRandomess in case of
+// unexpected size.
+var ErrRValueSize = errors.New("r-value must have 14 hex digits")
+
+// leastHalfTraceIDThresholdMask is the mask to use on the
+// least-significant half of the TraceID, i.e., bytes 8-15.
+// Because this is a 56 bit mask, the result after masking is
+// the unsigned value of bytes 9 through 15.
+//
+// This helps extract 56 bits of randomness from the second half of
+// the TraceID, as specified in https://www.w3.org/TR/trace-context-2/#randomness-of-trace-id
+const leastHalfTraceIDThresholdMask = MaxAdjustedCount - 1
+
+// Randomness may be derived from R-value or TraceID.
+//
+// Randomness contains 56 bits of randomness, derived in one of two ways, see:
+// https://www.w3.org/TR/trace-context-2/#randomness-of-trace-id
+type Randomness struct {
+	// unsigned is in the range [0, MaxAdjustedCount-1]
+	unsigned uint64
+}
+
+// TraceIDToRandomness returns randomness from a TraceID (assumes
+// the traceparent random flag was set).
+func TraceIDToRandomness(id pcommon.TraceID) Randomness {
+	// To get the 56 bits we want, take the second half of the trace ID,
+	leastHalf := binary.BigEndian.Uint64(id[8:])
+	return Randomness{
+		// Then apply the mask to get the least-significant 56 bits / 7 bytes.
+		// Equivalently stated: zero the most-significant 8 bits.
+		unsigned: leastHalf & leastHalfTraceIDThresholdMask,
+	}
+}
+
+// RValueToRandomness parses NumHexDigits hex bytes into a Randomness.
+func RValueToRandomness(s string) (Randomness, error) {
+	if len(s) != NumHexDigits {
+		return Randomness{}, ErrRValueSize
+	}
+
+	unsigned, err := strconv.ParseUint(s, hexBase, 64)
```

Comment:
This will parse both upper and lower case hex. The otep specifies lower case only (we talked about it and I lost that battle). Given that this code *generates* lower case, I don't mind being generous in parsing upper as well, but I wanted to note the discrepancy.

Reply:
:+1: I don't want to be forced to write a custom decoder, mainly. I'm not sure it matters, but if pressed, we can check for upper-case.